### PR TITLE
Isolate GCC variables.

### PR DESCRIPTION
**Pull request by marxin**

### CODE REVIEW

**Comment by marxin** [comment]
It requires the following counterpart:
https://gcc.gnu.org/g:ecb7b9c41347dba4087efe3c87d195e70297594c

File: texi2rst.py [hunk 1]
```diff
@@ -484,6 +484,22 @@ def postvisit_element(self, element, parent):
     return tree
 
 
+def fixup_text_variables(tree):
+    class TextVariableFixer(NoopVisitor):
+        REPLACEMENTS = (('$$VERSION_PACKAGE$$', '|package_version|'),
+                        ('$$version-GCC$$', '|gcc_version|',),
+                        ('$$BUGURL$$', '|bugurl|'))
```
Comment (davidmalcolm) [inline texi2rst.py:491]: Is the `|var|` syntax from .rst/sphinx, or something you've invented?

Comment (marxin) [reply texi2rst.py:491]: There are defined in `baseconf.py`, it uses standard `::replace` directive:
https://github.com/davidmalcolm/texi2rst/blob/46d828911d9c59fe96e6cc2105014c9003050e8e/scripts/templates/baseconf.py#L47-L51

**Comment by davidmalcolm** [comment]
> It requires the following counterpart:
> https://gcc.gnu.org/g:ecb7b9c41347dba4087efe3c87d195e70297594c

Looking at that patch:

Are the hunks like this deliberate?
```
-@ifset DEVELOPMENT
+ifsetDEVELOPMENT
 @emph{Warning:} This document, and the compiler it describes, are still
 under development.  While efforts are made to keep it up-to-date, it might
 not accurately reflect the status of the most recent GNU Fortran compiler.
-@end ifset
+endifset
```

Why not preserve the `@value{foo}` in the XML representation, and have texi2rst handle it, rather than patching it upfront?

**Comment by marxin** [comment]
> Are the hunks like this deliberate?

That's still on my TODO list.

> 
> ```
> -@ifset DEVELOPMENT
> +ifsetDEVELOPMENT
>  @emph{Warning:} This document, and the compiler it describes, are still
>  under development.  While efforts are made to keep it up-to-date, it might
>  not accurately reflect the status of the most recent GNU Fortran compiler.
> -@end ifset
> +endifset
> ```
> 
> Why not preserve the `@value{foo}` in the XML representation, and have texi2rst handle it, rather than patching it upfront?

The problem here is that `@ifset DEVELOPMENT` does not create any XML element. We'll need to wrap everything in between `ifset` and `endifset` in a element that will be Directive:

```
.. only:: Internal

  Content to be included only in the internal version of the document.
```

**Comment by davidmalcolm** [comment]
FWIW, I found this page helpful on this kind of thing:
  https://stackoverflow.com/questions/40689206/whats-the-difference-between-sphinxs-only-and-ifconfig-directives

"only" directive:
https://www.sphinx-doc.org/en/1.4.8/markup/misc.html#directive-only

How about synthesizing an XML element based on the ifset/endifset pairs, and moving the relevant parts of the DOM tree there.

Presumably the xml DOM tree currently looks something like:

```
TEXT containing "@ifset DEVELOPMENT"
other DOM nodes
TEXT containting "@endifset"
```

Could texi2xml handle `@ifset` and `@endifset` by creating elements, so we'd have:

```
TEXT up to the "@ifset DEVELOPMENT"
ELEMENT for ifset
other DOM nodes, as siblings
ELEMENT for endifset
TEXT after the "@endifset" clause
```

and then we could do a DOM maninpulation to turn it into:
```
TEXT up to the "@ifset DEVELOPMENT"
ELEMENT for ifset/endifset
  other DOM nodes, now as children
TEXT after the "@endifset" clause
```
and then the `ELEMENT for ifset/endifset` could become an `only` directive at the .rst level

**Comment by marxin** [comment]
All right, we can do an element wrapping at the XML level. What I problematic is that `md.texi` file contains quite many `@ifset INTERNALS` and `@ifclear INTERNALS` and one cannot build `gcc.xml` and `gccinternal.xml` with text replacement of the condition into `ifset INTERNALS`. If we don't convert them to Text, then they are not present in the created XML file.
The conditional build introduces many references and sections that clash.

I'm leaving that for now (and I reverted @ifset changes in `devel/sphinx` branch)...

**Pull request merged**